js/initcontext.go

```diff
@@ -110,7 +119,47 @@ func newBoundInitContext(base *InitContext, ctxPtr *context.Context, rt *goja.Ru
 		programs:          programs,
 		compatibilityMode: base.compatibilityMode,
 		logger:            base.logger,
+
+		shares:     base.shares,
+		sharesLock: base.sharesLock,
+	}
+}
+
+// XShare is a constructor returning a shareable read-only array (general objects coming)
+// indentified by the name and having their contents be whatever the call returns - again only
+// arrays currently supported.
+func (i *InitContext) XShare(name string, call goja.Callable) goja.Value {
+	i.sharesLock.Lock()
+	defer i.sharesLock.Unlock()
+	value, ok := i.shares[name]
+	if !ok { //nolint:nestif
+		gojaValue, err := call(goja.Undefined())
+		if err != nil {
+			common.Throw(i.runtime, err)
+		}
+		// TODO this can probably be better handled
+		if gojaValue.ExportType().Kind() != reflect.Slice {
+			common.Throw(i.runtime, errors.New("only arrays can be made into shared objects")) // TODO better error
+		}
+
+		// TODO this can probably be done better if we just iterate over the internal array, but ...
+		// that might be a bit harder given what currently goja provides
+		var tmpArr []interface{}
+		if err = i.runtime.ExportTo(gojaValue, &tmpArr); err != nil {
+			common.Throw(i.runtime, err)
+		}
+
+		arr := make([][]byte, len(tmpArr))
+		for index := range arr {
+			arr[index], err = json.Marshal(tmpArr[index])
```

Comment:
This is good enough for the first version, but I wonder if we couldn't figure out something better. For example, gob (or some other binary encoding/decoding) is probably faster. 

Or maybe no encoding and decoding is necessary at all, and we can share the raw Go objects and use [`Object.freeze()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze) to make it read-only and prevent data races? goja [seems to support it](https://github.com/dop251/goja/blob/6060b0671c09a09fbe8a33e29dfdbae33f85a640/builtin_object.go#L261-L287), and it even seems to work:
```js
export default function () {
    const obj = [1, 2, 3];
    obj[1] = 22;
    Object.freeze(obj); // comment this for no error
    console.log(obj);
    obj[2] = 33; // this throws a TypeError if frozen
    console.log(obj);
}
```

Even if it's not perfect and there's some obscure way to modify the object, we'll document that the shared data is read-only, so it'd be fine.

Reply:
My problem is that this freezes only the top object not any object that is property of the top object. Also, goja does sometimes have an internal state for even simple types that change without any locking or stuff ... because JS is single-threaded by definition. 

So IMO this will break horribly, but was one of the initial things I thought about ... also this means that I get an object from 1 VM and put it in another, without any kind of checking that it doesn't for example have functions ... that modify the original VMs properties or something ... as a whole I think while this approach will probably be "faster", unless goja has support for this it will probably be way too much work to actually properly check that things won't blow up, and any change to goja will likely necessitate refactoring. 

Maybe gob will be a better fit - maybe we should try it for a second iteration ... I mostly went directly for JSON as we already do this for the setupData ... and also ... JS(ON) ;)

Reply:
hmm didn't realize `Object.freeze()` only touches the first layer, but even that's easy to work around. Copying the public domain `deepFreeze()` from [here](https://stackoverflow.com/questions/34776846/how-to-freeze-nested-objects-in-javascript) (or, originally, [here](https://github.com/substack/deep-freeze/blob/master/index.js)) seems to work well enough:

```js
function deepFreeze(o) {
    Object.freeze(o);
    if (o === undefined) {
        return o;
    }

    Object.getOwnPropertyNames(o).forEach(function (prop) {
        if (o[prop] !== null
            && (typeof o[prop] === "object" || typeof o[prop] === "function")
            && !Object.isFrozen(o[prop])) {
            deepFreeze(o[prop]);
        }
    });

    return o;
};

export default function () {
    const obj = [1, [1, 2], { c: 3, d: 4 }];
    deepFreeze(obj); // comment this for no error
    console.log(JSON.stringify(obj));
    obj[1].push(3); // this throws a TypeError if frozen
    obj[2].f = 5; // and this also
    console.log(JSON.stringify(obj));
}
```

And we already have a JS wrapper anyway, why not go all the way? :sweat_smile:

Reply:
> Also, goja does sometimes have an internal state for even simple types that change without any locking or stuff ... because JS is single-threaded by definition.

So I don't think this is a good idea - we can explore it in a second iteration ;)

Reply:
> Also, goja does sometimes have an internal state for even simple types that change without any locking or stuff ... because JS is single-threaded by definition.

Sorry, I didn't address this properly. I know about the internal state, when you pass a goja `Runtime` some Go value, it wraps it in its own internal types. And that should be fine, since it doesn't touch the original value, as far as I know. At most it'll make a copy of a string to convert it to utf-16 or something like that...

So I think it's worth it to try using `Object.freeze()` (or, rather, `deepFreeze()`) even now, in the first iteration. It's going to be fine, even if it doesn't work at 100% and some JS guru can get around it and still modify something. They are doing it on their own risk of a data race... Besides, it's not like we haven't had issues with transforming objects to JSON and back before - we've had different key names and Go strangeness with that as well. And we do it for `setup()` because we have to ship the data between instances in the cloud. Here the requirements are much different.

Reply:
I don't completely understand everything Dmitry says, but it seems like this won't be feasible... :disappointed: 

One benefit of recursively freezing objects instead of copying them (i.e. marshaling and unmarshaling), besides improved performance, would have been the better UX. With the current solution, each runtime will get its own copy of the array item, and would seeming be able to make changes in it. But if it accesses the same array item the next time, the changes it made on the previous iteration will be gone, since it'd receive a fresh copy of the data... Not the end of the world, but unless we document it very well, someone is liable to spend a bit of time pulling their hair. So, maybe it's worth recursively `Object.freeze()` things even when we're copying the data on the Go side?